src/tools/cardos5-tool.c

```diff
@@ -0,0 +1,1245 @@
+
+	memset(&apdu, 0, sizeof(apdu));
+	apdu.cse = SC_APDU_CASE_3_SHORT;
+	apdu.cla = data[0];
+	apdu.ins = data[1];
+	apdu.p1 = data[2];
+	apdu.p2 = data[3];
+	if (data_len > 4) {
+		/* Inconsistent LC or LE != 0. */
+		if (data[4] != data_len - 4)
+			errx(1, "%s: bogus apdu in %s", __func__, apdu_path);
+		if (data_len > 5) {
+			apdu.lc = data[4];
+			apdu.data = &data[5];
+			apdu.datalen = data[4];
+		}
+	}
+
+	if (apdu.cla != 0x84 || apdu.ins != 0x24)
+		errx(1, "%s: bogus apdu in file %s", __func__, apdu_path);
+
+	r = sc_transmit_apdu(card, &apdu);
+	if (r != SC_SUCCESS)
+		errx(1, "%s: sc_transmit_apdu: %s", __func__, sc_strerror(r));
+
+	if (apdu.sw1 != 0x90 || apdu.sw2 != 0x00)
+		errx(1, "%s: command failed: %02x %02x", __func__, apdu.sw1,
+		    apdu.sw2);
+
+	free(data);
+}
+
+#ifdef ENABLE_OPENSSL
+/*
+ * Calculate the Message Authentication Code (MAC) of "data" according to AES
+ * CMAC 128 ECB using "key" as the cryptographic key. The result is stored in
+ * "mac", which is allocated with *mac_len bytes.
+ *
+ * This function fails if AES_BLOCK_SIZE is not 16.
+ */
+int
+aes_cmac(const uint8_t *data, size_t data_len, const uint8_t *key,
+    size_t key_len, uint8_t **mac, size_t *mac_len)
+{
+#ifdef HAVE_OPENSSL_CMAC_H
+	CMAC_CTX	*ctx;
+	uint8_t		 tmp[AES_BLOCK_SIZE];
+	size_t		 tmp_len;
+
+	ctx = CMAC_CTX_new();
+	if (ctx == NULL)
+		return -1;
+
+	if (CMAC_Init(ctx, key, key_len, EVP_aes_128_cbc(), NULL) == 0 ||
+	    CMAC_Update(ctx, data, data_len) == 0 ||
+	    CMAC_Final(ctx, tmp, &tmp_len) == 0 ||
+	    tmp_len != AES_BLOCK_SIZE ||
+	    tmp_len != 16) {
+		CMAC_CTX_free(ctx);
+		return -1;
+	}
+
+	CMAC_CTX_free(ctx);
+
+	/*
+	 * CardOS 5.0 User's Manual, section 2.6.2 (pg. 89) states: "From the
+	 * 16 bytes returned by AES-CMAC computation, the most significant 8
+	 * bytes are used for secure messaging purposes."
+	 */
+
+	*mac = calloc(8, sizeof(char));
+	if (*mac == NULL)
+		return -1;
+
+	memcpy(*mac, tmp, 8);
+	*mac_len = 8;
+
+	return 0;
+#else
+	warnx("%s: AES-CMAC support missing in OpenSSL", __func__);
+	return -1;
+#endif
+}
+
+/*
+ * Perform AES CBC as described in CardOS 5.0 User's Manual. The relevant
+ * sections and figures are: 2.6.2 (pg. 89), 2.6.2-2 (pg. 91), and 4.1.1 in the
+ * Packages & Release Notes (PRN).
+ *
+ * We receive a system key (which is the StartKey) and use the invariant seed
+ * value to derive an encryption key from it. This derivation process is
+ * underdocumented and therefore unclear; the only hint in the manual being
+ * "SMx4H_ENC_KEY = ENC_SYSTEMKEY(SEED)". It does not state which
+ * initialisation vector (IV) should be used in this calculation.
+ *
+ * Once in possession of the encryption key, figure 2.6.2-2 (referenced above)
+ * suggests that we need to encrypt 16 zero bytes to get the IV to be used when
+ * encrypting the data payload.
+ */
+int
+aes_cbc(const uint8_t *data, size_t data_len, const uint8_t *syskey,
+    size_t syskey_len, uint8_t **out, size_t *out_len)
+{
```

Comment:
AES-CBC is already implemented mulitple times in `card-epass2003.c`, `card-piv.c` and `sc-hsm-tool.c`. This should definitely be tied together...